File: solver.go
package main

import (
	"fmt"
	"sync"
	"time"
)

// AddWord adds a word to the trie
func AddWord(trie *Trie, words *sync.Map, word string) {
	if len(word) > 2 && trie.SearchWord(word) {
		words.Store(word, true)
	}
}

// ValidRecurse checks if the next move is valid on the board
func ValidRecurse(newLocation [2]int, visitedSoFar map[[2]int]bool, board [][]string) bool {
	// Check if we're within the bounds of the board:
	if newLocation[0] >= len(board) || newLocation[1] >= len(board[0]) ||
		newLocation[0] < 0 || newLocation[1] < 0 {
		return false
	}

	// Check if we've been to this location before
	return !visitedSoFar[newLocation]
}

// SolveRecursive is a recursive function to find words on the board
func SolveRecursive(trie *Trie, board [][]string, visitedSoFar map[[2]int]bool, location [2]int, wordSoFar string, words *sync.Map) {
	visitedSoFar[location] = true
	AddWord(trie, words, wordSoFar)

	if trie.isLeaf(wordSoFar) {
		delete(visitedSoFar, location)
		return
	}

	// Go in all directions
	directions := [8][2]int{{-1, 1}, {0, 1}, {1, 1}, {-1, 0}, {1, 0}, {-1, -1}, {0, -1}, {1, -1}}
	for _, direction := range directions {
		newLocation := [2]int{location[0] + direction[0], location[1] + direction[1]}
		if ValidRecurse(newLocation, visitedSoFar, board) {
			visitedSoFar[newLocation] = true
			newWordSoFar := wordSoFar + board[newLocation[0]][newLocation[1]]
			SolveRecursive(trie, board, visitedSoFar, newLocation, newWordSoFar, words)
			delete(visitedSoFar, newLocation)
		}
	}
}

// SolveThread is a goroutine for solving the board in a threaded manner
func SolveThread(root *Trie, letters [][]string, row, col int, words *sync.Map, wg *sync.WaitGroup) {
	defer wg.Done()
	visitedSoFar := make(map[[2]int]bool)
	location := [2]int{row, col}
	SolveRecursive(root, letters, visitedSoFar, location, letters[row][col], words)
}

// Solve solves the board and returns the set of found words
func Solve(letters [][]string) *sync.Map {
	start := time.Now()
	words := new(sync.Map)
	root := NewTrie()
	root.populate()
	var wg sync.WaitGroup

	for row := 0; row < len(letters); row++ {
		for col := 0; col < len(letters[0]); col++ {
			wg.Add(1)
			go SolveThread(root, letters, row, col, words, &wg)
		}
	}

	wg.Wait()
	elapsed := time.Since(start)
	fmt.Printf("Solving took: %s\n", elapsed)
	return words
}

func main() {
	var numLines int
	fmt.Scan(&numLines)
	var board [][]string
	for i := 0; i < numLines; i++ {
		var newChar string
		var currLine []string
		for j := 0; j < numLines; j++ {
			fmt.Scan(&newChar)
			currLine = append(currLine, newChar)
		}
		board = append(board, currLine)
	}
	words := Solve(board)

	word_list := []string{}
	words.Range(func(key, value interface{}) bool {
		word_list = append(word_list, key.(string))
		return true
	})
	fmt.Printf("Found %d Words: ", len(word_list))
	for _, element := range word_list {
		fmt.Printf(element + ", ")
	}
}

File: solver.cpp
#include <chrono>
#include <sstream>
#include "trie.h"
#include <unordered_set>
#include <vector>
using namespace std;

int directions[8][2] = {{-1, 1}, {0, 1}, {1, 1}, {-1, 0}, {1, 0}, {-1, -1}, {0, -1}, {1, -1}};

void addWord(struct TrieNode *trie, unordered_set<string>* words, string word) {
    if (word.length() > 2 && search(trie, word)) {
        (*words).insert(word);
    }
}

bool validRecurse(int newLocation[2], vector<int*> visitedSoFar, vector<vector<string>> board) {
    // Check if we're within the bounds of the board:
    if (newLocation[0] >= board.size() || newLocation[1] >= board[0].size() || 
        newLocation[0] < 0 || newLocation[1] < 0) {
        return false;
    }
    // Check if we've been to this location before
    for (int i = 0; i < visitedSoFar.size(); i++) {
        if (visitedSoFar[i][0] == newLocation[0] & visitedSoFar[i][1] == newLocation[1]) {
            return false;
        }
    }
    return true;
}

void solveRecursive(struct TrieNode *trie, vector<vector<string>> board, vector<int*> visitedSoFar, int location[2], string wordSoFar, unordered_set<string>* words) {
    visitedSoFar.push_back(location);
    addWord(trie, words, wordSoFar);
    if (isLeaf(trie, wordSoFar)) {
        visitedSoFar.pop_back();
        return;
    }
    // go in all directions
    for (int i = 0; i < 8; i++) {
        int newLocation[2] = {location[0] + directions[i][0], location[1] + directions[i][1]};
        if (validRecurse(newLocation, visitedSoFar, board)) {
            visitedSoFar.push_back(newLocation);
            string newWordSoFar = wordSoFar + board[newLocation[0]][newLocation[1]];
            solveRecursive(trie, board, visitedSoFar, newLocation, newWordSoFar, words);
            visitedSoFar.pop_back();
        }
    }
}

unordered_set<string>* solve(vector<vector<string>> letters) {
    auto start = chrono::system_clock::now();
    unordered_set<string>* words = new unordered_set<string>;
    struct TrieNode *root = create_dictionary();
    vector<int*> visitedSoFar;
    for (int row = 0; row < letters.size(); row++) {
        for (int col = 0; col < letters[0].size(); col++) {
            int location[2] = {row, col};
            solveRecursive(root, letters, visitedSoFar, location, letters[row][col], words);
        }
    }
    auto elapsed = chrono::duration_cast<std::chrono::seconds>(start.time_since_epoch());
    cout << "Solving took: " << elapsed.count() << endl;
    return words;
}

int main() {
    vector<vector<string>> board;
    string line;
    cin >> line;
    int num_lines = stoi(line);
    for (int i = 0; i < num_lines; i++) {       
        string new_char;
        vector<string> curr_line;
        for(int j = 0; j < num_lines; j++) {
            cin >> new_char;
            curr_line.push_back(new_char);
        }
        board.push_back(curr_line);
    }
    unordered_set<string>* words = solve(board);
    cout << "Found " << (*words).size() << " Words: ";
    for (string word : *words) {
        cout << word << ", ";
    }
    return 0;
}